crates/ra_hir_ty/src/display.rs

```diff
@@ -6,46 +6,98 @@ use crate::{
     db::HirDatabase, utils::generics, ApplicationTy, CallableDef, FnSig, GenericPredicate,
     Obligation, ProjectionTy, Substs, TraitRef, Ty, TypeCtor,
 };
-use hir_def::{generics::TypeParamProvenance, AdtId, AssocContainerId, Lookup};
+use hir_def::{
+    find_path, generics::TypeParamProvenance, item_scope::ItemInNs, AdtId, AssocContainerId,
+    Lookup, ModuleId,
+};
 use hir_expand::name::Name;
 
-pub struct HirFormatter<'a, 'b> {
+pub struct HirFormatter<'a, R> {
     pub db: &'a dyn HirDatabase,
-    fmt: &'a mut fmt::Formatter<'b>,
+    fmt: &'a mut dyn fmt::Write,
```

Comment:
Trait object vs type parameter?

Reply:
So far, all calls to `HirDisplay::{display, display_truncated}` are either followed by `.to_string()` or are inserted into a format string (if I haven't missed anything), so maybe we could also just remove the intermediate `HirDisplayWrapper` and just have these methods return a `String`? Would there be any drawbacks?

Reply:
Calling `to_string` or inserting into a format string are exactly the uses of the wrapper, but in the latter case it's wasteful to print into a separate string first :slightly_smiling_face: